### PR TITLE
GH-36594: [C++] Don't use MSVC_VERSION to determin -fms-compatibility-version

### LINKED ISSUE #36594: [C++][Gandiva] MSVC_VERSION not handled correctly
### Describe the bug, including details regarding any error messages, version, and platform.

The following logic is deficient:
https://github.com/apache/arrow/blob/05620b186f2d27fd9d28dab010c2fa58e67c0ce7/cpp/src/gandiva/precompiled/CMakeLists.txt#L33-L47

It will never reach the `elseif` or the `else` (as it just decides between <1930, yes or no), and will pass an empty `FMS_COMPATIBILITY` for VS2022.

Also worth noting, [elsewhere](https://github.com/apache/arrow/blob/05620b186f2d27fd9d28dab010c2fa58e67c0ce7/cpp/cmake_modules/SetupCxxFlags.cmake#L171-L173) `MSVC_VERSION` uses 19, and not 19xx. This is also wrong according to the CMake [docs](https://cmake.org/cmake/help/latest/variable/MSVC_VERSION.html).

### Component(s)

C++ - Gandiva

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

`MSVC_VERSION` isn't used correctly. 

### What changes are included in this PR?

* Remove MSVC 2013 or earlier check because it reached EOL
* Don't change `-fms-compatibility-version` depending on `MSVC_VERSION`
   * Always use the lowest supported Visual Studio version for it

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* Closes: #36594

### CODE REVIEW

File: .github/workflows/cpp.yml [hunk 1]
```diff
@@ -301,15 +299,15 @@ jobs:
...
+          key: cpp-ccache-windows-${{ matrix.version }}-${{ env.CACHE_VERSION }}-${{ hashFiles('cpp/**') }}
+          restore-keys: cpp-ccache-windows-${{ matrix.version }}-${{ env.CACHE_VERSION }}-
         env:
           # We can invalidate the current cache by updating this.
           CACHE_VERSION: "2022-09-13"
       - name: Build
         shell: cmd
         run: |
-          call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
+          call "C:\Program Files (x86)\Microsoft Visual Studio\*\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
```
Comment (kou) [inline .github/workflows/cpp.yml:310]:
```suggestion
          call "C:\Program Files (x86)\Microsoft Visual Studio\${{ matrix.version }}\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64
```

File: .github/workflows/cpp.yml [hunk 2]
```diff
@@ -219,18 +219,16 @@ jobs:
...
       fail-fast: false
       matrix:
-        os:
-          - windows-2019
-        include:
-          - os: windows-2019
-            name: Windows 2019
+        version:
+          - 2019
+          - 2022
```
[the remark this answers is not in the document]
Comment (kou) [reply .github/workflows/cpp.yml:231]: OK.
I'll add `verify-rc-source-windows-2022` and remove this after we confirm that this changes work with MSVC 2022 here.

**Comment by pitrou** [comment]
@kou What's the status on this PR?

**Comment by kou** [comment]
I tried confirming this change on GitHub Actions but I abandoned because LLVM binaries for Windows that can be used as library. LLVM provides the official LLVM binaries at https://github.com/llvm/llvm-project/releases/tag/llvmorg-16.0.0 but they only include tools such as `clang.exe`.

So we can't test these changes in CI for now.

**Comment by pitrou** [comment]
@kou Do you want to rebase and merge this nevertheless?

**Comment by kou** [comment]
I revisited this.

We can use LLVM as a library with conda. So we can test this case with `conda-win-x64-cpu-py3`.

I think that we don't need to use different `-fms-compatibility-version` depending on the building Visual Studio version. I think that we can always use the lowest supported Visual Studio version for it. (Visual Studio 16 2019 now. Because Visual Studio 15 2017's "lifecycle stage" is "Extended" not "Mainstream": https://learn.microsoft.com/en-us/visualstudio/productinfo/vs-servicing#older-versions-of-visual-studio )

**Comment by kou** [comment]
@github-actions crossbow submit conda-win-x64-cpu-py3

**Comment by kou** [comment]
@github-actions crossbow submit conda-win-x64-cpu-py3

**Comment by kou** [comment]
@pitrou Do you have any more comments? Can we merge this?

**Review by pitrou — APPROVED** [review]

**Pull request merged**